bundles/org.openhab.binding.goveelan/src/main/java/org/openhab/binding/goveelan/internal/GoveeLanDiscoveryService.java

```diff
@@ -0,0 +1,226 @@
+                        logger.trace("Govee Device Response: {}", response);
+
+                        Map<String, Object> properties = getDeviceProperties(response);
+                        final String sku = properties.get(GoveeLanConfiguration.DEVICETYPE).toString();
+                        final String skuLabel = "discovery.goveelan.goveeLight." + sku;
+                        String productName = i18nProvider.getText(bundleContext.getBundle(), skuLabel, sku,
+                                Locale.getDefault());
+                        properties.put(GoveeLanConfiguration.PRODUCTNAME, (productName != null) ? productName : sku);
+                        ThingUID thingUid = new ThingUID(GoveeLanBindingConstants.THING_TYPE_LIGHT,
+                                properties.get(GoveeLanConfiguration.MAC_ADDRESS).toString().replace(":", "_"));
+
+                        DiscoveryResultBuilder discoveryResult = DiscoveryResultBuilder.create(thingUid)
+                                .withProperties(properties)
+                                .withRepresentationProperty(GoveeLanConfiguration.MAC_ADDRESS).withLabel(
+                                        "Govee " + productName + " " + properties.get(GoveeLanConfiguration.DEVICETYPE)
+                                                + " (" + properties.get(GoveeLanConfiguration.IPADDRESS) + ")");
+
+                        thingDiscovered(discoveryResult.build());
+                    } while (true); // left by SocketTimeoutException
+                } catch (SocketTimeoutException ste) {
+                    // done with scanning
+                } catch (IOException e) {
+                    logger.warn("Discovery with IO exception: {}", e.getMessage());
+                }
+            });
+        } catch (UnknownHostException e) {
+            logger.warn("Discovery failed: {}", e.getMessage());
+        } catch (InterruptedException e) {
+        } finally {
+            discoveryActive = false;
+        }
+    }
+
+    public Map<String, Object> getDeviceProperties(String response) {
+        Gson gson = new Gson();
+
+        DiscoveryMessage message = gson.fromJson(response, DiscoveryMessage.class);
+        String ipAddress = message.msg().data().ip();
+        String deviceType = message.msg().data().sku();
+        String macAddress = message.msg().data().device();
+
+        Map<String, Object> properties = new HashMap<>(3);
+        properties.put(GoveeLanConfiguration.IPADDRESS, ipAddress);
+        properties.put(GoveeLanConfiguration.DEVICETYPE, deviceType);
+        properties.put(GoveeLanConfiguration.MAC_ADDRESS, macAddress);
+
+        return properties;
+    }
+
+    private List<NetworkInterface> getLocalNetworkInterfaces() {
+        List<NetworkInterface> result = new LinkedList<>();
+        try {
+            for (NetworkInterface networkInterface : Collections.list(NetworkInterface.getNetworkInterfaces())) {
```

Comment:
You could also iterate directly through the enumerator - see for example:
https://github.com/openhab/openhab-addons/blob/01dc99c6b6da7956daad718e7da7ad19380b21bf/bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/discovery/DanfossAirUnitDiscoveryService.java#L92-L107

Reply:
major refactoring